Fix GDP error when using a disjunction with nested disjunctions as a target

emma58 opened the pull request:
## Fixes #1933 

## Summary/Motivation:

GDP transformations usually get lucky because they transform objects in declaration order, and Disjuncts have to be declared before the Disjunctions they appear in. However, you can mess up this order in the `targets` list. In particular, in the referenced issue, you can give a Disjunction which contains nested Disjunctions as a target. This bypasses the depth-first-search transformation order, so GDP fails to transform the nested stuff and then blames you.

## Changes proposed in this PR:
- Preprocess targets so that Disjuncts always appear before the Disjunctions that contain them in the list.
- Remove tests for the active Disjunction error, as that is now unreachable, I think
- Be more careful about the argument to the transformation: must be Block-like so that we are sure we catch this in targets

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

jsiirola commented on pyomo/gdp/plugins/bigm.py:
```diff
@@ -231,9 +237,32 @@ def _apply_to_impl(self, instance, **kwds):
+                    if t.is_indexed():
+                        for disjunction in t.values():
+                            for disj in disjunction.disjuncts:
+                                print("adding Disjunct %s" % disj.name)
```
Is this debugging? If this is expected output, please convert to using the logger.

jsiirola commented on pyomo/gdp/plugins/bigm.py:
```diff
@@ -231,9 +237,32 @@ def _apply_to_impl(self, instance, **kwds):
+                                preprocessed_targets.append(disj)
+                    else:
+                        for disj in t.disjuncts:
+                            print("adding Disjunct %s" % disj.name)
```
Is this debugging? If this is expected output, please convert to using the logger.

jsiirola commented on pyomo/gdp/plugins/bigm.py:
```diff
@@ -231,9 +237,32 @@ def _apply_to_impl(self, instance, **kwds):
+                # now we are safe to put the disjunction, and if the target was
+                # anything else, then we don't need to worry because disjuncts
+                # are declared before disjunctions they appear in
+                print("adding target %s of type %s" % (t.name, t.ctype))
```
Is this debugging? If this is expected output, please convert to using the logger.

jsiirola commented on pyomo/gdp/plugins/hull.py:
```diff
@@ -230,13 +230,41 @@ def _apply_to(self, instance, **kwds):
         targets = self._config.targets
         if targets is None:
             targets = ( instance, )
+        else:
+            # we need to preprocess targets to make sure that if there are any
+            # disjunctions in targets that their disjuncts appear before them in
+            # the list.
+            preprocessed_targets = []
+            for t in targets:
+                if t.ctype is Disjunction:
+                    if t.is_indexed():
+                        for disjunction in t.values():
+                            for disj in disjunction.disjuncts:
+                                print("adding Disjunct %s" % disj.name)
+                                preprocessed_targets.append(disj)
+                    else:
+                        for disj in t.disjuncts:
+                            print("adding Disjunct %s" % disj.name)
+                            preprocessed_targets.append(disj)
+                # now we are safe to put the disjunction, and if the target was
+                # anything else, then we don't need to worry because disjuncts
+                # are declared before disjunctions they appear in
+                print("adding target %s of type %s" % (t.name, t.ctype))
+                preprocessed_targets.append(t)
+            targets = preprocessed_targets
```
This seems identical to the implementation in bigm.  Should we create a utility method to prevent repeating code?

emma58 replied:
Yes... This is not the only place we have this problem, but I may as well not add to it.

jsiirola reviewed (changes requested):
In addition to the specific comments, one general one:  what happens if given:
```python
m = ConcreteModel()
m.d1 = Disjunct()
m.d2 = Disjunct()
m.disj = Disjunction(expr=[m.d1, m.d2])
```
the user specifies something like:
```python
TransformationFactory('gdp.bigm').apply_to(m, targets=[m.d1, m.disj])
```
or
```python
TransformationFactory('gdp.bigm').apply_to(m, targets=[m, m.disj])
```
(that is, the same block appears "twice" in the targets list)?

emma58 commented on pyomo/gdp/plugins/hull.py:
```diff
@@ -230,13 +230,41 @@ def _apply_to(self, instance, **kwds):
+                # now we are safe to put the disjunction, and if the target was
+                # anything else, then we don't need to worry because disjuncts
+                # are declared before disjunctions they appear in
+                print("adding target %s of type %s" % (t.name, t.ctype))
```
Ooof, yes, it's debugging. Sorry, I apparently forgot to take *all* of them out...

emma58 commented on pyomo/gdp/plugins/bigm.py:
```diff
@@ -213,6 +213,12 @@ def _apply_to(self, instance, **kwds):
             self.used_args.clear()
 
     def _apply_to_impl(self, instance, **kwds):
+        if not instance.ctype in (Block, Disjunct):
```
Perhaps? I'm not sure either. But I guess there is no reason we shouldn't go along with something's Blockiness. I think that means line 261 should change also, yes?

emma58 commented:
@jsiirola regarding your question, it's fine if targets appear multiple times in the list because we just skip over already-transformed components (because there's nothing to stop me from transforming an inner disjunction with bigm and then the outer with hull or something like that). I added tests to that effect though... 

I addressed everything except your question about accepting all subclasses of Block, which I'm still thinking about. I *think* we should since it might save us from some of the incompatibilities with other packages. And it seems like the expected behavior, right?
